### PR TITLE
Implement post-quantum block level access control

### PR DESCRIPTION
**Pull request by ianopolous**
This adds a capability based scheme for authorising retrieval of a block. To retrieve a block you need a Block Access Token (BAT) which is 32 random bytes. To retrieve a block you use an applicable BAT to generate an S3 V4 signature and include that in the request over bitswap (or send the BAT directly to the Peergos API). The signature incorporates the nodeId of the sender in place of the domain name for the target for S3. This allows us to broadcast the same signature to the DHT and each node can verify it was sent from the actual signing node, and protects against replay attacks. 

Cryptree nodes store a list of BATs under the cbor key "bats" in the top level. They also store a nested BAT for each linked raw fragment. Each user has a mirror BAT which can be used to mirror their entire block set. A server can also have a server wide BAT which can be used to retrieve any block on the instance. More details on BATs are here: https://github.com/Peergos/Peergos/issues/862

For files, subsequent chunk BATs are derived using the stream secret in the same way as the map keys. This allows us to keep the O(1) seeking in huge files (just local hashing). 

Closes: https://github.com/Peergos/Peergos/issues/862

### CODE REVIEW

File: src/peergos/server/storage/S3BlockStorage.java [hunk 1]
```diff
@@ -94,14 +100,23 @@ private Multihash keyToHash(String key) {
...
-    public CompletableFuture<List<PresignedUrl>> authReads(List<Multihash> blocks) {
+    public CompletableFuture<List<PresignedUrl>> authReads(List<MirrorCap> blocks) {
         if (blocks.size() > 50)
             throw new IllegalStateException("Too many reads to auth!");
         List<PresignedUrl> res = new ArrayList<>();
 
-        for (Multihash block : blocks) {
-            String s3Key = hashToKey(block);
-            res.add(S3Request.preSignGet(s3Key, Optional.of(600), ZonedDateTime.now(), host, region, accessKeyId, secretKey));
+        // retrieve all blocks and verify BATs in aprallel
```
Comment (kevodwyer) [inline src/peergos/server/storage/S3BlockStorage.java:108]: spelling of parallel

File: src/peergos/shared/storage/auth/Bat.java [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+            throw new IllegalStateException("Cannot use identity multihash in S3 signatures!");
+        S3Request req = new S3Request("GET", sourceNode.toBase58(), "api/v0/block/get?arg=" + block.toBase58(), S3Request.UNSIGNED,
+                Optional.of(expirySeconds), false, true,
+                Collections.emptyMap(), Collections.emptyMap(), batId.toBase58(), "eu-central-1", datetime);
+        return S3Request.computeSignature(req, encodeSecret(), h)
+                .thenApply(signature -> new BlockAuth(ArrayOps.hexToBytes(signature), expirySeconds, datetime, batId));
+    }
+
+    public static Bat deriveFromRawBlock(byte[] block) {
+        // TODO need to analyse the cryptographic properties of this construction, might need to use hmac_sha256 instead
```
Comment (kevodwyer) [inline src/peergos/shared/storage/auth/Bat.java:85]: this still a todo?

**Review by kevodwyer — CHANGES_REQUESTED** [review]
I didn't mark all the places with the word Cave. I would suggest dropping it.

File: src/peergos/shared/storage/auth/Bat.java (hunk 2 above)
Comment (ianopolous) [reply src/peergos/shared/storage/auth/Bat.java:85]: yep.

File: src/peergos/server/Main.java [hunk 3]
```diff
@@ -770,7 +803,9 @@ public static boolean migrate(Args a) {
             }
             System.out.println("Migrating user from node " + currentStorageNodeId + " to " + newStorageNodeId);
             List<UserPublicKeyLink> newChain = Migrate.buildMigrationChain(existing, newStorageNodeId, user.signer.secret);
-            user.network.coreNode.migrateUser(username, newChain, currentStorageNodeId).join();
+            List<BatWithId> mirrorBats = user.network.batCave.getUserBats(username, user.signer).join();
+            BatWithId current = mirrorBats.get(mirrorBats.size() - 1);
```
[the remark this answers is not in the document]
Comment (ianopolous) [reply src/peergos/server/Main.java]: done

**Review by kevodwyer — APPROVED** [review]

**Pull request merged**